staging/src/k8s.io/legacy-cloud-providers/aws/aws.go

```diff
@@ -3860,21 +3860,33 @@ func (c *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, apiS
 		}
 	} else {
 		klog.V(4).Infof("service %v does not need custom health checks", apiService.Name)
-		// We only configure a TCP health-check on the first port
-		var tcpHealthCheckPort int32
-		for _, listener := range listeners {
-			if listener.InstancePort == nil {
-				continue
-			}
-			tcpHealthCheckPort = int32(*listener.InstancePort)
-			break
-		}
 		annotationProtocol := strings.ToLower(annotations[ServiceAnnotationLoadBalancerBEProtocol])
 		var hcProtocol string
+		var tcpHealthCheckPort int32
 		if annotationProtocol == "https" || annotationProtocol == "ssl" {
 			hcProtocol = "SSL"
+			for _, listener := range listeners {
+				if listener.InstancePort == nil {
+					continue
+				}
+				protocol := listener.Protocol
+				if *protocol == "HTTPS" || *protocol == "SSL" {
+					tcpHealthCheckPort = int32(*listener.InstancePort)
+					break
+				}
+				if tcpHealthCheckPort == 0 {
```

Comment:
Shouldn't this check be outside the loop?

Reply:
Ah yes, you would be correct. I'll fix that up.